### PR TITLE
Lower Matmul operation to Krnl dialect

### PR DESCRIPTION
**Pull request by tungld**
TODO
- [x] Add lowering tests
- [x] 1-D x 1-D
- [x] 1-D x N-D, N >=2, and vice versa.

Note: there are no ONNX backend tests for 1-D cases.

~~Functions `createIterateOperandPack` and `insertKrnlOps` are picked up from #61 in order to make the code succinct.~~

### CODE REVIEW

File: src/pass/lower_frontend_to_krnl.cpp [hunk 1]
```diff
@@ -227,6 +227,126 @@ getLoopIVsForBroadcasting(Location loc, ConversionPatternRewriter &rewriter,
   return newLoopIVs;
 }
 
+// Create an iterate operand pack for KrnlIterateOp
+KrnlIterateOperandPack createIterateOperandPack(
```
Comment (doru1004) [inline src/pass/lower_frontend_to_krnl.cpp]: This code is un multiple PRs. We should not have code that is duplicated across PRs.

File: src/pass/lower_frontend_to_krnl.cpp [hunk 2]
```diff
@@ -227,6 +227,126 @@ getLoopIVsForBroadcasting(Location loc, ConversionPatternRewriter &rewriter,
...
+      pack.pushConstantBound(memRefShape[i]);
+    }
+  }
+
+  return pack;
+}
+
+// Insert a block of KrnlDefineLoopsOp, KrnlOptimizeLoopsOp and KrnlIterateOp.
+std::tuple<KrnlDefineLoopsOp, KrnlOptimizeLoopsOp, KrnlIterateOp>
+insertKrnlOps(Location loc, PatternRewriter &rewriter,
```
Comment (doru1004) [inline src/pass/lower_frontend_to_krnl.cpp]: Similar here.

File: src/pass/lower_frontend_to_krnl.cpp [hunk 3]
```diff
@@ -1314,14 +1717,13 @@ struct ONNXGemmOpLowering : public ConversionPattern {
     std::vector<Value> outerLoops, optimizedOuterLoops;
     outerLoops.reserve(2);
     optimizedOuterLoops.reserve(2);
-    for (int i = 0; i < 2; ++i) {
+    for (int64_t i = 0; i < 2; ++i) {
```
Comment (doru1004) [inline src/pass/lower_frontend_to_krnl.cpp]: There is no need for these counters to be int64_t. Can we change them back to int ?

File: src/dialect/onnx/onnx_ops.cpp [hunk 4]
```diff
@@ -394,7 +394,7 @@ void ONNXMatMulOp::inferShapes() {
         lhsShape[0] != rhsShape[rhsRank - 2])
       emitError("Attempt to multiply incompatible matrices.");
 
-    for (int i = 0; i < rhsRank - 2; ++i)
+    for (int64_t i = 0; i < rhsRank - 2; ++i)
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: Change to int.

File: src/dialect/onnx/onnx_ops.cpp [hunk 5]
```diff
@@ -412,7 +412,7 @@ void ONNXMatMulOp::inferShapes() {
         lhsShape[lhsRank - 1] != rhsShape[0])
       emitError("Attempt to multiply incompatible matrices.");
 
-    for (int i = 0; i < lhsRank - 2; ++i)
+    for (int64_t i = 0; i < lhsRank - 2; ++i)
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: int

File: src/dialect/onnx/onnx_ops.cpp [hunk 6]
```diff
@@ -426,7 +426,7 @@ void ONNXMatMulOp::inferShapes() {
         lhsShape[lhsRank - 1] != rhsShape[0])
       emitError("Attempt to multiply incompatible matrices.");
 
-    for (int i = 0; i < lhsRank - 1; ++i)
+    for (int64_t i = 0; i < lhsRank - 1; ++i)
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: int

File: src/dialect/onnx/onnx_ops.cpp [hunk 7]
```diff
@@ -440,7 +440,7 @@ void ONNXMatMulOp::inferShapes() {
         lhsShape[1] != rhsShape[rhsRank - 2])
       emitError("Attempt to multiply incompatible matrices.");
 
-    for (int i = 0; i < rhsRank - 2; ++i)
+    for (int64_t i = 0; i < rhsRank - 2; ++i)
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: int

File: src/dialect/onnx/onnx_ops.cpp [hunk 8]
```diff
@@ -458,10 +458,10 @@ void ONNXMatMulOp::inferShapes() {
 
     // Check and perform broadcasting for the shapes.
     SmallVector<int64_t, 2> lhsBcastShape;
-    for (int i = 0; i < lhsRank - 2; ++i)
+    for (int64_t i = 0; i < lhsRank - 2; ++i)
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: int

File: src/dialect/onnx/onnx_ops.cpp [hunk 9]
```diff
@@ -458,10 +458,10 @@ void ONNXMatMulOp::inferShapes() {
...
       lhsBcastShape.emplace_back(lhsShape[i]);
     SmallVector<int64_t, 2> rhsBcastShape;
-    for (int i = 0; i < rhsRank - 2; ++i)
+    for (int64_t i = 0; i < rhsRank - 2; ++i)
```
Comment (doru1004) [inline src/dialect/onnx/onnx_ops.cpp]: int

File: src/pass/lower_frontend_to_krnl.cpp [hunk 10]
```diff
@@ -1234,6 +1362,279 @@ struct ONNXReshapeOpLowering : public ConversionPattern {
...
+    if (hasAllConstantDimensions(memRefType))
+      alloc = insertAllocAndDealloc(memRefType, loc, rewriter, insertDealloc);
+    else {
+      SmallVector<Value, 4> allocOperands;
+      if (AShape.size() >= 2 && BShape.size() >= 2) {
+        // Both arguments are N-D, N >= 2
+        // (s1 x s2 x... x sK x M x K) MATMUL (K x N)
+        // =>
+        // (s1 x s2 x... x sK x M x N)
+        for (int64_t i = 0; i < memRefShape.size() - 2; ++i) {
```
Comment (doru1004) [inline src/pass/lower_frontend_to_krnl.cpp]: int

**Comment by doru1004** [comment]
In general it looks good. I think there are too many int64_t iteration variables even where they don't need to be. Can you revise the need for int64_t for some of the loop counters please? I've commented on some of the ones I spotted I may have missed some.

File: src/dialect/onnx/onnx_ops.cpp (hunk 4 above)
Comment (tjingrant) [reply src/dialect/onnx/onnx_ops.cpp]: @doru1004 int64_t is actually appropriate here; the problem here is that `rhsRank` should also be an int64_t type. And `rhsRank` should ideally be obtained using `getRank` method provided in standard types, which returns an int64_t integer.

c.f.:
https://github.com/llvm/llvm-project/blob/3ccaac3cdd8f1e5b19e2da04be2ebbfc1fb9aa32/mlir/lib/IR/StandardTypes.cpp#L136

Comment (doru1004) [reply src/dialect/onnx/onnx_ops.cpp]: Ah yes you are right. it's actually being used directly. There are other cases where the iteration variable is only used an index.

**Comment by tungld** [comment]
Doru, I revised the code. Please take a look again. Thanks!

File: src/pass/lower_frontend_to_krnl.cpp [hunk 11]
```diff
@@ -1210,6 +1218,336 @@ struct ONNXReshapeOpLowering : public ConversionPattern {
...
+      rewriter.create<KrnlReturnLoopsOp>(loc, reduceLoops);
+
+      // Insert instructions into the reduction KrnlIterateOp.
+      Block &reduceIterationBlock = reduceIterateOp.bodyRegion().front();
+      rewriter.setInsertionPointToStart(&reduceIterationBlock);
+
+      // Induction variables
+      SmallVector<Value, 4> loopKIVs, loopOuterMKIVs, loopOuterKNIVs;
+      // K
+      loopKIVs.emplace_back(reduceIterationBlock.getArguments()[0]);
```
Comment (tjingrant) [inline src/pass/lower_frontend_to_krnl.cpp]: nit: getArgument(0).

File: src/pass/lower_frontend_to_krnl.cpp [hunk 12]
```diff
@@ -1210,6 +1218,336 @@ struct ONNXReshapeOpLowering : public ConversionPattern {
...
+        auto AB = rewriter.create<MulIOp>(loc, loadedA, loadedB);
+        auto accumulated = rewriter.create<AddIOp>(loc, loadedY, AB);
+        rewriter.create<StoreOp>(loc, accumulated, alloc, zeroIndex);
+      } else if (elementType.isa<FloatType>()) {
+        auto AB = rewriter.create<MulFOp>(loc, loadedA, loadedB);
+        auto accumulated = rewriter.create<AddFOp>(loc, loadedY, AB);
+        rewriter.create<StoreOp>(loc, accumulated, alloc, zeroIndex);
+      }
+    } else {
+      // No scalar matrix multiplication.
```
Comment (tjingrant) [inline src/pass/lower_frontend_to_krnl.cpp]: Seems like a perfect place to use llvm_unreachable.

File: src/pass/lower_frontend_to_krnl.cpp [hunk 13]
```diff
@@ -1210,6 +1218,336 @@ struct ONNXReshapeOpLowering : public ConversionPattern {
...
+    // A value zero
+    Value zero;
+    if (elementType.isa<IntegerType>()) {
+      zero = rewriter.create<ConstantOp>(
+          loc, IntegerAttr::get(memRefType.getElementType(), 0));
+    } else if (elementType.isa<FloatType>()) {
+      zero = rewriter.create<ConstantOp>(
+          loc, FloatAttr::get(memRefType.getElementType(), 0));
+    } else {
+      emitError(loc, "unsupported element type");
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/pass/lower_frontend_to_krnl.cpp]: With `return emitError`, we got errors:
```
error: no viable conversion from returned value of type 'mlir::InFlightDiagnostic' to function return type
      'mlir::PatternMatchResult' (aka 'Optional<std::unique_ptr<PatternState> >')
```

It seemed return types mismatched.

[the remark this answers is not in the document]
Comment (doru1004) [reply src/pass/lower_frontend_to_krnl.cpp]: I suspect emitError needs to be fixed so for now we can rely it will be fixed at some point.

**Comment by tjingrant** [comment]
@doru1004 I'm okay with this patch now, do you have any more comments?

**Comment by doru1004** [comment]
> @doru1004 I'm okay with this patch now, do you have any more comments?

I still see int64_t loop variables where they shouldn't be.

**Comment by tungld** [comment]
@doru1004 I changed `int64_t` back to `int`. The reason I used `int64_t` was that MemRefType's shape is `ArrayRef<int64_t>` (c.f. https://github.com/llvm/llvm-project/blob/master/mlir/include/mlir/IR/StandardTypes.h#L208).

I see that, in your `addDimensionToPack`, `int` indices are used to obtain shape dimensions, I am suspect that it is unsafe.

File: src/dialect/onnx/onnx_ops.cpp (hunk 4 above)
Comment (doru1004) [reply src/dialect/onnx/onnx_ops.cpp]: I take back my previous comment, please use int here. The reason is you are using "I" as an index into the shape vector.

Comment (doru1004) [reply src/dialect/onnx/onnx_ops.cpp]: This is applicable to all my other comments that are still outstanding.

**Comment by doru1004** [comment]
> @doru1004 I changed `int64_t` back to `int`. The reason I used `int64_t` was that MemRefType's shape is `ArrayRef<int64_t>` (c.f. https://github.com/llvm/llvm-project/blob/master/mlir/include/mlir/IR/StandardTypes.h#L459).
> 
> I see that, in your `addDimensionToPack`, `int` indices are used to obtain shape dimensions, I am suspect that it is unsafe. => (updated) I am sorry, it seems I misunderstood indices can be `int`. Please ignore this comment.

There are some outstanding comments still.

**Comment by tungld** [comment]
> There are some outstanding comments still.

Fixed them. Thanks!

**Pull request merged**